osgar/lib/quaternion.py

```diff
@@ -36,9 +36,16 @@ def rotate_vector(vector, quaternion):
     return multiply(part1, con)[:-1]
 
 def euler_zyx(quaternion):
+    # https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles
+    # assumption that quaternion is normalized
     x0, y0, z0, w0 = quaternion
+    sqr_size = x0*x0 + y0*y0 + z0*z0 + w0*w0
+    if abs(sqr_size - 1.0) > 0.00001:
+        k = math.sqrt(sqr_size)
+        x0, y0, z0, w0 = x0/k, y0/k, z0/k, w0/k
     ax =  math.atan2(2*(w0*x0+y0*z0), 1-2*(x0*x0+y0*y0))
-    ay =  math.asin(2*(w0*y0-z0*x0))
+    v = min(1.0, max(-1.0, 2*(w0*y0-z0*x0)))
+    ay =  math.asin(v)
```

Comment:
I don't understand why these changes are needed. When does normalized quaternion hits the min/max conditions?

Reply:
This was the original fix and I left it there as I have seen in my life that floats sometimes underflow, like sqrt of negative zero. Also on the original reference:
https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles
they take care (rather) of the limits too:
```
// pitch (y-axis rotation)
    double sinp = 2 * (q.w * q.y - q.z * q.x);
    if (std::abs(sinp) >= 1)
        angles.pitch = std::copysign(M_PI / 2, sinp); // use 90 degrees if out of range
    else
        angles.pitch = std::asin(sinp);
```

Reply:
I am reading http://www.euclideanspace.com/maths/geometry/rotations/conversions/quaternionToEuler/index.htm in the hope to find an answer (there are some singularities).